hivemind/p2p/servicer.py

```diff
@@ -0,0 +1,97 @@
+from typing import Any, Callable, Optional, Union
+
+from hivemind.p2p.p2p_daemon import P2P, P2PContext
+from hivemind.p2p.p2p_daemon_bindings.datastructures import PeerID
+
+
+@dataclass
+class RPCHandler:
+    method_name: str
+    handle_name: str
+    request_type: type
+    response_type: type
+
+
+class StubBase:
+    """
+    Base class for P2P RPC stubs. The interface mimicks gRPC stubs.
+
+    Servicer derives stub classes for particular services (e.g. DHT, averager, etc.) from StubBase,
+    adding the necessary rpc_* methods. Calls to these methods are translated to calls to the remote peer.
+    """
+
+    def __init__(self, p2p: P2P, peer: PeerID):
+        self.p2p = p2p
+        self.peer = peer
+
+
+class Servicer:
+    """
+    Base class for P2P RPC servicers (e.g. DHT, averager, MoE server). The interface mimicks gRPC servicers.
+
+    - ``add_p2p_handlers(self, p2p)`` registers all rpc_* methods of the derived class as P2P unary handlers, allowing
+      other peers to call them. It uses type annotations for the ``request`` parameter and the return value
+      to infer protobufs the methods operate with.
+
+    - ``get_stub(self, p2p, peer)`` creates a stub with all rpc_* methods. Calls to the stub methods are translated
+      to calls to the remote peer.
+    """
+
+    def __init__(self):
+        class_name = self.__class__.__name__
+
+        self._rpc_handlers = []
+        for method_name, method in self.__class__.__dict__.items():
+            if method_name.startswith('rpc_') and callable(method):
+                handle_name = f'{class_name}.{method_name}'
+
+                hints = method.__annotations__
+                try:
+                    request_type = self._hint_to_type(hints['request'])
+                    response_type = self._hint_to_type(hints['return'])
+                except (KeyError, ValueError):
+                    raise ValueError(f'{handle_name} is expected to have type annotations like `dht_pb2.FindRequest` '
+                                     f'(a type from the hivemind.proto module) for the `request` parameter '
+                                     f'and the return value')
+
+                self._rpc_handlers.append(RPCHandler(method_name, handle_name, request_type, response_type))
+
+        self._stub_type = type(f'{class_name}Stub', (StubBase,),
+                               {handler.method_name: self._make_rpc_caller(handler)
+                                for handler in self._rpc_handlers})
+
+    @staticmethod
+    def _make_rpc_caller(handler: RPCHandler):
+        async def caller(stub: StubBase, request: handler.request_type,
+                         timeout: Optional[float] = None) -> handler.response_type:
+            return await asyncio.wait_for(
+                stub.p2p.call_unary_handler(stub.peer, handler.handle_name, request, handler.response_type),
+                timeout=timeout)
+
+        caller.__name__ = handler.method_name
+        return caller
```

Comment:
Copying the entire docstring would be incorrect since their parameters differ (the servicer method accepts `request: protobuf, context: P2PContext` and the stub method accepts `request: proto, timeout: Optional[float] = None`).